odoo/addons/base/models/res_users.py

```diff
@@ -786,6 +786,61 @@ def change_password(self, old_passwd, new_passwd):
         # use self.env.user here, because it has uid=SUPERUSER_ID
         return self.env.user.write({'password': new_passwd})
 
+    @api.model
+    def _deactivate_portal_user(self):
+        """Try to remove the current portal user.
+
+        If it failed, archive the user instead and remove sensitive information.
+
+        This is used to give the opportunity to portal users to de-activate their accounts.
+        Indeed, as the portal users can easily create accounts, they sometimes will wish
+        it removed because they don't use this Odoo portal anymore.
+
+        Before this feature, they would have to contact the website or the support to get
+        their account removed, which could be tedious.
+        """
+        if not self.env.user.has_group('base.group_portal'):
+            raise UserError(_('Only the portal users can deactivate their accounts.'))
+
+        ip = request.httprequest.environ['REMOTE_ADDR'] if request else 'n/a'
+        _logger.info('Account deletion asked for "%s" (#%i) from %s', self.env.user.login, self.env.uid, ip)
+
+        log_user = f'User #{self.env.uid}:'
+
+        # A user can not self-deactivate, switch to super user
+        user = self.env.user.with_user(SUPERUSER_ID)
+
+        try:
+            self._cr.execute('SAVEPOINT remove_portal_user')
+            partner = user.partner_id
+            user.unlink()
+            partner.unlink()
+
+            _logger.info('%s User and partner deleted from the database.', log_user)
+        except Exception:
+            # Can not unlink the users, maybe because of constraints on other models
+            # remove sensitive information and archive the record
+            self._cr.execute('ROLLBACK TO SAVEPOINT remove_portal_user')
+            _logger.info('%s Deletion of the user failed, archive it instead.', log_user)
+
+            _logger.info('%s Change the password.', log_user)
+            user.password = os.urandom(32).hex()
+
+            deleted_name = f'__deleted_user_{user.id}_{time.time()}'
+
+            _logger.info('%s Archive the user and remove sensitive information.', log_user)
+
+            user.write({
+                'name': deleted_name,
+                'login': deleted_name,
+            })
+
+            user.action_archive()
+            _logger.info('%s Archive the partner.', log_user)
```

Comment:
This is a sensitive operation, so every log line should ideally include the full context, considering that it can be interspersed in the middle of 3 millions other lines. It must be easy to grep, and include the necessary traceability info :)

e.g. each log msg could include `"User deletion request by <login> #<uid>"` and `"via IP <ip>"`